File: LeetCode/ProblemSet/Volume (3800-3899)/LeetCode3867/LeetCode3867-Solution-01.py
from typing import *
from math import gcd

class Solution:
    def gcdSum(self, nums: list[int]) -> int:
        n, maxNum = len(nums), 0
        for i, num in enumerate(nums):
            maxNum = max(maxNum, num)
            nums[i] = gcd(maxNum, num)

        nums.sort()

        return sum(gcd(nums[i], nums[n - i - 1]) for i in range(0, n // 2))

File: LeetCode/ProblemSet/Volume (3800-3899)/LeetCode3867/LeetCode3867-Solution-01.cpp
#include <bits/stdc++.h>

class Solution {
public:
    long long gcdSum(std::vector<int>& nums) {
        const int n = (int) nums.size();

        int maxNum = 0;
        for (int i = 0; i < n; i++) {
            maxNum = std::max(maxNum, nums[i]);
            nums[i] = std::gcd(nums[i], maxNum);
        }

        std::ranges::sort(nums);

        long long sum = 0;
        int i = 0, j = n - 1;
        while (i < j) {
            sum += std::gcd(nums[i], nums[j]);
            i++, j--;
        }

        return sum;
    }
};